### DISCUSSION TITLE
Solving IPS Display Problems

### DISCUSSION THREAD

**Discussion by Electrotechnique** [ElectroTechnique/electrotechnique.github.io#4] [Hardware Issues]
The display used in TSynth is a low-cost IPS 80x160 colour display with SPI interface, on a small PCB. These occasionally have a slightly different display fitted to the PCB, which will cause problems like the following:

![IPS_Problem](https://user-images.githubusercontent.com/50934180/105619053-0c241880-5e32-11eb-8f41-5f002b90aed1.jpg)

The displays can be differentiated (sometimes!) by the colour of the tab on the protective film, although this isn't reliable or convenient when you're ordering from a foreign supplier. The solution is to modify the code that initialises the display (in ST7735Display.h). The following lines can be changed:

```
  tft.initR(INITR_GREENTAB); //INITR_BLACKTAB 
  tft.invertDisplay(true); // false
```
Occasionally you may find the display is offset, which requires some adjustment to code in ST7735_t3.h. See Survival Hacking's video at about 31:30 - it's in Italian but is easy to follow if you don't speak it: [Survival Hacking TSynth Part 2](https://www.youtube.com/watch?v=MusKKTRkITU)

Any other problems with displays, post here and we'll endeavour to fix them.

**Comment by luisfcorreia** [comment]
Hey, it's my display :)

I should have opened a discussion for this

**Reply by alfrede** [reply]
my display is quite dark even with fw 2.10

**Comment by seomradubh** [comment]
This solution worked for me, thanks for that, great work!

**Comment by don-recardo** [comment]
I too have the dreaded slanting text . I have read what two lines  I need to change and i I read somewhere I have to Compile with 'Serial + MIDI + Audio . and I havent a clue what that means  

Does someone have the latest V2.31 as a hex file but with the lines altered to fix my display . I know how to load a hex file to the teensy but as for making a hex file , I am totaly lost . If I could scrounge a copy that would be great 

Cheers 
Don

**Comment by don-recardo** [comment]
If anyone has the same sloping text display here is the fixed hex file which Iain from 
ElectroTechnique has kindly allowed me to post

Don

https://www.dropbox.com/s/uix2ky5z6r2d95f/TSynth-fixed-display.ino.hex?dl=0

**Comment by rpopp-fni-stl-com** [comment]
Hello, I posted on the facebook builders group, but thought I would also post here.  I just completed my build on a 4.1 board, and while everything works, my display is rotated and mirrored (reads right to left).  I can fix the rotation, but I cannot for the life of me figure out the mirroring.  The display I bought has the correct pinout but it has a _blue_ tab, not green, black or red.  

FWIW, that display is here:  https://ebay.to/3b1jyz2

Any help would be immensely appreciated

**Comment by Electrotechnique** [comment]
Is there anything here that might help?:

[https://github.com/Bodmer/TFT_eSPI/issues/143](https://github.com/Bodmer/TFT_eSPI/issues/143)

**Reply by rpopp-fni-stl-com** [reply]
Sorry to be a pain, but you'll have to bear with me a little bit.  What exactly am I looking for here?

**Reply by rpopp-fni-stl-com** [reply]
So, after much experimentation, I was finally able to get my display oriented correctly.  That said, if I format a 16G sd card (on linux), as FAT 32 LBA, and then copy the entirety of the contents of patches over, my Tsynth still says NO SD at boot.  Is FAT 32 LBA not supported?

**Reply by rpopp-fni-stl-com** [reply]
Disregard the above.  I was able to format a FAT16 sd card and load samples.

**Reply by Electrotechnique** [reply]
Great, can you tell us what you did to sort out the mirroring issue?

**Comment by marcel-licence** [comment]
Do you have still issues with the Display? It looks like adafruit does not support all variants. I forked the library and added a change to use the display without issues. https://github.com/marcel-licence/Adafruit-ST7735-Library I am not sure if that might fix your observed issue

**Comment by cestballot** [comment]
> The display used in TSynth is a low-cost IPS 80x160 colour display with SPI interface, on a small PCB. These occasionally have a slightly different display fitted to the PCB, which will cause problems like the following:
> 
> ![IPS_Problem](https://user-images.githubusercontent.com/50934180/105619053-0c241880-5e32-11eb-8f41-5f002b90aed1.jpg)
> 
> The displays can be differentiated (sometimes!) by the colour of the tab on the protective film, although this isn't reliable or convenient when you're ordering from a foreign supplier. The solution is to modify the code that initialises the display (in ST7735Display.h). The following lines can be changed:
> 
> ```
>   tft.initR(INITR_GREENTAB); //INITR_BLACKTAB 
>   tft.invertDisplay(true); // false
> ```
> 
> Occasionally you may find the display is offset, which requires some adjustment to code in ST7735_t3.h. See Survival Hacking's video at about 31:30 - it's in Italian but is easy to follow if you don't speak it: [Survival Hacking TSynth Part 2](https://www.youtube.com/watch?v=MusKKTRkITU)
> 
> Any other problems with displays, post here and we'll endeavour to fix them.

Hello, I have the same issu, like this picture. I have a 2.10 firmware with the good correction, but is it possible to have the new 2.32 firmware with it in the .hex ( without to build a custom firmware) ? 

Best regards

**Reply by Electrotechnique** [reply]
I've compiled a V2.32 firmware with the changes: https://github.com/ElectroTechnique/TSynth-Teensy4.1/tree/master/Firmware

Can you download the **entire** repository to get the file. I have not tested it as I don't have the alternative displays.

**Reply by cestballot** [reply]
Thanks, it looks good. Display works and now I'll test basic fonctions. Thanks for your reactivity.

**Comment by pathein** [comment]
Hi, I have updated to v2.33 recently and having the display issue.

May I know if anyone having v2.33 hex with fixed display to share. Thanks in advance

**Reply by Electrotechnique** [reply]
Hello can you try this please:
[https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynthV2_34_INVDISP.zip](https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynthV2_34_INVDISP.zip)

**Comment by pathein** [comment]
Thanks so much for the file, will try again.

**Comment by pathein** [comment]
![V2 34](https://user-images.githubusercontent.com/22908290/174348949-7c85eaca-56d1-4889-976d-f0aab5f6bedd.jpg)

Thanks for the file, It works good. For the background, its now in white and brighter.

**Reply by Electrotechnique** [reply]
Ok, the display needs inverting. Try this:
[https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynthV2_34DISP.zip](https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynthV2_34DISP.zip)

**Reply by pathein** [reply]
Perfect screen now, back to the black background! Thanks so much for the file

**Comment by MonCalamari** [comment]
Hi,

im am building a tsynth, i have 3 different displays, they all have the same issue. When i use the 2.33 firmware the display is mirrored  upside down. Since i have tried 3 different displays i like to fix the Problem in Software, but i dont know how. Can anybody help me?

**Reply by Electrotechnique** [reply]
Hello, can you provide a picture so that I can decide what to try changing.

**Reply by MonCalamari** [reply]
Sure, here is a picture with the Firmware provided on github:

![PXL_20221124_114433989 MP](https://user-images.githubusercontent.com/32960993/203776977-5fbfbeea-9bb5-48c1-ac4f-05f088246220.jpg)

**Reply by Electrotechnique** [reply]
What colour is the tab on the protective film, that might give a clue.

**Comment by luisfcorreia** [comment]
is that really that deterministic?

In this case Blue Label

**Reply by Electrotechnique** [reply]
Apparently yes, there are the following choices in ST7735_t3.h:

#define INITR_GREENTAB 0x0
#define INITR_REDTAB   0x1
#define INITR_BLACKTAB 0x2

#define INITR_18GREENTAB    INITR_GREENTAB
#define INITR_18REDTAB      INITR_REDTAB
#define INITR_18BLACKTAB    INITR_BLACKTAB
#define INITR_144GREENTAB   0x1
#define INITR_144GREENTAB_OFFSET   0x4
#define INITR_MINI160x80  0x05


No blue though. I'll have a look at the current version and have a play.

**Reply by MonCalamari** [reply]
Hi, yes, blue label. It seems this is the only Version i can get in germany right now. I have asked some shops, all of them have the blue label dsiplays.

**Reply by MonCalamari** [reply]
Please let me know if there is anything i can do to solve the problem.

**Reply by Electrotechnique** [reply]
I might have to give you several firmwares to try. I'll get back to you soon.

**Reply by MonCalamari** [reply]
Ok, thanks for your support!

**Reply by Electrotechnique** [reply]
Yes, someone reported this problem [earlier in this thread](https://github.com/ElectroTechnique/electrotechnique.github.io/discussions/4#discussioncomment-1518278), but didn't respond to a request to share the fix. I might use the version of ST7735_t3 supplied with the current Teensyduino 1.57 rather than the 1.48 version that's currently used locally in the TSynth code..

**Comment by Electrotechnique** [comment]
Can you try this and report on what happens please, it's initialising with `tft.initB();`  

[https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynthV2_33_initB.zip](https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynthV2_33_initB.zip)

**Reply by MonCalamari** [reply]
Hi, yes, tried this, no change.

![PXL_20221125_090942892 NIGHT](https://user-images.githubusercontent.com/32960993/203943420-a77aab46-d9c8-4cbc-b108-e6a01a79044b.jpg)

**Reply by MonCalamari** [reply]
I can get a Display with a green Tab from a friend. Is this case intresting and you want to fix it (Maybe for other Builders)? Otherwise i can change the display to the green one.

**Reply by Electrotechnique** [reply]
Yes it would be good to know the fix, as this will probably arise again. If you want to swap the display and get finished, that's fine. I'll carry on trying. I find it odd that initB() makes no difference - except the offset error on the left is gone. On my green tab display, I get nothing, so it is doing something. 

It looks like the display is simply flipping the x axis and working from bottom upwards, and it's possibly rgb order and not bgr - altering the colours.

**Reply by MonCalamari** [reply]
No, i can wait and i am happy to help in some way

**Reply by Electrotechnique** [reply]
Can you try this:
[https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynth_V2_33_MADCTL_MV.zip](https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynth_V2_33_MADCTL_MV.zip)

If it works, there may still be an offset, but that's easier to fix.

**Comment by MonCalamari** [comment]
Nice, it works now, the offset seems fine too. Thank you!

![PXL_20221125_144510962 MP](https://user-images.githubusercontent.com/32960993/204008616-af9e4335-2b90-401a-abc5-f9bf6de26ae8.jpg)

**Reply by Electrotechnique** [reply]
Good. For anyone else with the same problem, change line 714 in the ST7735_t3.cpp in the TSynth folder:

from
` writedata_last(MADCTL_MX | MADCTL_MV | MADCTL_BGR);`
to
 ` writedata_last(MADCTL_MV | MADCTL_RGB);`

It's a quick solution.

**Reply by Electrotechnique** [reply]
Ok, try this one:
[https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynth_V2_33_MADCTL_Offset.zip](https://github.com/ElectroTechnique/electrotechnique.github.io/blob/master/TSynth_V2_33_MADCTL_Offset.zip)

It might need a bit more. It's line 468, _rowstart = 0;

**Reply by MonCalamari** [reply]
Yes, this works totaly fine. Ok, now i know where to change the offset an it is documented here so other builders can try it by themself. Thank you again, i hope my audioboard arrives today so i can test the tsynth.

**Comment by cantevolution** [comment]
Hello, during the installation of Electrotechnic TSynth, I'm facing two issues. 
The first one is that my PC doesn't allow me to program directly through the USB client port. I've tried to summarize the problem through screenshots (THIS IS ERROR MESSAGE "device reported a problem and was discontinued. (Code 43)USB device descriptor request failed."). It doesn't recognize the serial for programming. I proceeded by programming directly from the T-Synth processor's USB. 
The more significant issue is that I can't properly see the display; it remains completely black but active. I attempted to reprogram following instructions in this link "[https://github.com/.../electrotechnique.../discussions/64](https://github.com/ElectroTechnique/electrotechnique.github.io/discussions/64?fbclid=IwAR1EXwLzaYAB7NMDwjRfaTORqaYL-5mn4MIWpWCW7Twa8NUm61OSjgSIR6M)", using the latest software version for the Chinese display (this: [https://www.aliexpress.us/item/3256801351207015.html...](https://www.aliexpress.us/item/3256801351207015.html?spm=a2g0s.9042311.0.0.30fc4c4diWOkAr&gatewayAdapt=ita2usa4itemAdapt&fbclid=IwAR0zJFX2kDit1Kkos2TH59fXsOMODYWE40sHJS0AnxxpTYCeoIP_64bbNts) ), but the problem persists. It might be an installation-related problem. How can I resolve this or check if the board is functioning correctly? 
I've never been able to start it successfully. 
Thank you.
![WhatsApp Image 2023-11-20 at 23 59 21](https://github.com/ElectroTechnique/electrotechnique.github.io/assets/151585504/f3d69601-f479-4bde-8035-650df2c9e135)
![WhatsApp Image 2023-11-20 at 23 59 22](https://github.com/ElectroTechnique/electrotechnique.github.io/assets/151585504/d7ef0917-5bfc-45fa-98c3-5ec594629ff4)
![WhatsApp Image 2023-11-21 at 00 00 50](https://github.com/ElectroTechnique/electrotechnique.github.io/assets/151585504/f0b4d5e5-2a10-4a3f-949b-67081bb4c8c6)

**Comment by Electrotechnique** [comment]
Try to program the Teensy with a simple "Blink" example program. Check all solder joints.

**Reply by cantevolution** [reply]
SOLVED! NO CONNECTIONs WITH THE BOARD AND THE DISPLAY. I USED OTHER DISPLAY AND CABLES BECAUSE SOME TRACKS DIDN'T WORK WELL!
![WhatsApp Image 2023-11-21 at 19 30 22](https://github.com/ElectroTechnique/electrotechnique.github.io/assets/151585504/de30f6da-7d8e-43b9-b4b5-491aac7e7915)
